### ISSUE TITLE
WebRtcEndpoint (libnice) leaks socket file descriptors

### ISSUE THREAD

**Issue by omgold** [Kurento/bugtracker#460]
## Prerequisites

* [x] I have read the SUPPORT document
  

* [x] I have checked the Troubleshooting Guide
  

* [x] I have tested with the latest version of Kurento
  

## Issue description

After a lifecycle of the of a webrtcendpoint there are file descriptors on the media server process left, refering to network ports (2 of each UDP and TCP). The ports are in the range specified in BaseRtpEndpoint.conf.ini.

## Context

I'm using kms for Apache Openmeetings. I noticed the problem when audio/video stopped working after a few sessions, due to kms refusing to create new channels when it ran out of file descriptors (having around 900 open).

## How to reproduce?

Don't know how to build a minimal reproducer as I'm not familiar enough with gst. Could try to build one with some advice.

## Expected & current behavior

Ports/FDs should be cleanup up, so they won't run out.

## Environment info

- Kurento and Openmeetings are running on the same host. Openmeetings is exposed to the outside world through an apache reverse proxy (with websocket forwarding)
- STUN/TURN is not used. Kurento seems to be able to use end-to-end UDP anyway, despite both client and server being behind a NAT.

See the attached debug log for the lifecycle of one webrpcendpoint (enabling and disabling camera once in openmeetings). To me there are no obvious problems. Apparently all created gst elements are finalized afterwards.

### INFO: Kurento Media Server

* Kurento version: 6.13.0
* Server OS: Arch
* Installation method:
  - [x] Built from sources
    

### INFO: Application Server

* Language: Java
* Kurento Client version: 6.12.0

### INFO: End-user clients

* Device(s): PC
* OS(es): Linux (Arch/Ubuntu)
* Browser(s): FIrefox 75

### Version details

```
lsb_release -a
<not applicable>

kurento-media-server -v
Kurento Media Server version: 6.13.0

Found modules:
	'core' version 6.13.0
	'elements' version 6.13.0
	'filters' version 6.13.0

pacman -Q|grep -Pi 'kurento|kms-|gst-|nice'
gst-plugins-bad 1.16.2-7
gst-plugins-bad-libs 1.16.2-7
gst-plugins-base 1.16.2-1
gst-plugins-base-libs 1.16.2-1
gst-plugins-good 1.16.2-3
kms-core 6.13.0-2
kms-elements 6.13.0-1
kms-filters 6.13.0-2
kms-jsoncpp 1.6.4-1
kms-jsonrpc 6.13.0-1
kurento-media-server 6.13.0-1
libnice 0.1.16-1
```

**Comment by omgold** [comment]
[webrtc-lifecycle.log](https://github.com/Kurento/bugtracker/files/4523623/webrtc-lifecycle.log)

**Comment by j1elo** [comment]
Confirmed. We are at least leaking socket file descriptors from **libnice**.

This is the list of FDs used by KMS after running the "hello-world" tutorial twice, then closing the app and waiting for KMS to run its garbage collector:

```
$ lsof -a \
    -p "$(pgrep --newest --full "kurento-media-server")" \
    -d '^cwd,^err,^ltx,^mem,^mmap,^pd,^rtd,^txt'

COMMAND    PID USER   FD      TYPE             DEVICE SIZE/OFF    NODE NAME
kurento-m 3909 juan    0u      CHR              136,0      0t0       3 /dev/pts/0
kurento-m 3909 juan    1u      CHR              136,0      0t0       3 /dev/pts/0
kurento-m 3909 juan    2u      CHR              136,0      0t0       3 /dev/pts/0
kurento-m 3909 juan    3u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan    4u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan    5u  a_inode               0,13        0   12082 [eventpoll]
kurento-m 3909 juan    6u  a_inode               0,13        0   12082 [timerfd]
kurento-m 3909 juan    7u     IPv6            1235260      0t0     TCP *:8888 (LISTEN)
kurento-m 3909 juan    8r      CHR                1,9      0t0      11 /dev/urandom
kurento-m 3909 juan   10u  netlink                         0t0 1235255 KOBJECT_UEVENT
kurento-m 3909 juan   11r     FIFO               0,12      0t0 1235256 pipe
kurento-m 3909 juan   12w     FIFO               0,12      0t0 1235256 pipe
kurento-m 3909 juan   13r     FIFO               0,12      0t0 1235259 pipe
kurento-m 3909 juan   14w     FIFO               0,12      0t0 1235259 pipe
kurento-m 3909 juan   15u  a_inode               0,13        0   12082 [timerfd]
kurento-m 3909 juan   18u     unix 0x0000000000000000      0t0 1235511 type=STREAM
kurento-m 3909 juan   19u     unix 0x0000000000000000      0t0 1235512 type=STREAM
kurento-m 3909 juan   20u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   21u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   22u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   23u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   24u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   25u     IPv4            1237381      0t0     UDP jpc-1.home:33645 
kurento-m 3909 juan   26u     IPv4            1235666      0t0     TCP jpc-1.home:18212 (LISTEN)
kurento-m 3909 juan   27u     IPv4            1237388      0t0     UDP jpc-1.home:36467 
kurento-m 3909 juan   28u     IPv4            1237391      0t0     TCP jpc-1.home:4342 (LISTEN)
kurento-m 3909 juan   29u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   30u  a_inode               0,13        0   12082 [eventpoll]
kurento-m 3909 juan   31u  a_inode               0,13        0   12082 [timerfd]
kurento-m 3909 juan   32u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   33u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   34u  a_inode               0,13        0   12082 [eventpoll]
kurento-m 3909 juan   35u  a_inode               0,13        0   12082 [timerfd]
kurento-m 3909 juan   36u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   37u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   38u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   39u     IPv4            1240582      0t0     UDP jpc-1.home:8669 
kurento-m 3909 juan   40u     IPv4            1239706      0t0     TCP jpc-1.home:36295 (LISTEN)
kurento-m 3909 juan   41u     IPv4            1241522      0t0     UDP jpc-1.home:52829 
kurento-m 3909 juan   42u     IPv4            1240621      0t0     TCP jpc-1.home:5214 (LISTEN)
```

This pattern repeats for each time a new WebRtcEndpoint is created:

```
kurento-m 3909 juan   20u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   21u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   22u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   23u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   24u  a_inode               0,13        0   12082 [eventfd]
kurento-m 3909 juan   25u     IPv4            1237381      0t0     UDP jpc-1.home:33645 
kurento-m 3909 juan   26u     IPv4            1235666      0t0     TCP jpc-1.home:18212 (LISTEN)
kurento-m 3909 juan   27u     IPv4            1237388      0t0     UDP jpc-1.home:36467 
kurento-m 3909 juan   28u     IPv4            1237391      0t0     TCP jpc-1.home:4342 (LISTEN)
```

here I tested with 2 WebRtcEndpoints, so that this kind of block appears 2 times.

Now the UDP and TCP ports coincide with those chosen by libnice to gather local candidates:

```
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint0> [IceCandidateFound] local: 'candidate:1 1 UDP 2015363327 192.168.1.12 33645 typ host', stream_id: 1, component_id: 0
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint0> [IceCandidateFound] local: 'candidate:3 1 TCP 1010827519 192.168.1.12 18212 typ host tcptype passive', stream_id: 1, component_id: 0
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint0> [IceCandidateFound] local: 'candidate:1 2 UDP 2015363326 192.168.1.12 36467 typ host', stream_id: 1, component_id: 1
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint0> [IceCandidateFound] local: 'candidate:3 2 TCP 1010827518 192.168.1.12 4342 typ host tcptype passive', stream_id: 1, component_id: 1

INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint1> [IceCandidateFound] local: 'candidate:1 1 UDP 2015363327 192.168.1.12 8669 typ host', stream_id: 1, component_id: 0
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint1> [IceCandidateFound] local: 'candidate:3 1 TCP 1010827519 192.168.1.12 36295 typ host tcptype passive', stream_id: 1, component_id: 0
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint1> [IceCandidateFound] local: 'candidate:1 2 UDP 2015363326 192.168.1.12 52829 typ host', stream_id: 1, component_id: 1
INFO webrtcendpoint kmswebrtcendpoint.c:116:on_ice_candidate:<kmswebrtcendpoint1> [IceCandidateFound] local: 'candidate:3 2 TCP 1010827518 192.168.1.12 5214 typ host tcptype passive', stream_id: 1, component_id: 1
```

It is then safe to assume that either we have a bug in that libnice is not properly told to release its resources, or else it is a bug in libnice itself.

**Comment by omgold** [comment]
Did some analysis and found a workaround:

What happens is that kmswebrtcendpoint calls kms_ice_nice_agent_remove_stream() followed kms_ice_nice_agent_finalize().

kms_ice_nice_agent_remove_stream() calls nice_agent_remove_stream() which does not directly close the stream but istalls an async handler to do that (call chain within nice: nice_agent_remove_stream() -> refresh_prune_stream_async() -> on_refresh_removed() -> agent_timeout_add_with_context_internal() ). The handler function on_stream_refreshes_pruned() is supposed to finally call nice_stream_close().

Unfortunately before that happens the nice_agent object is destroyed in kms_ice_nice_agent_finalize(), which prevents the callback handler from being executed.

What makes the cleanup work for me, is to explicitly call g_main_context_iteration once in kms_ice_nice_agent_finalize() before destorying the nice_agent:

```
diff --git a/src/gst-plugins/webrtcendpoint/kmsiceniceagent.c b/src/gst-plugins/webrtcendpoint/kmsiceniceagent.c
index 3b1b10d..64c1a13 100644
--- a/src/gst-plugins/webrtcendpoint/kmsiceniceagent.c
+++ b/src/gst-plugins/webrtcendpoint/kmsiceniceagent.c
@@ -274,6 +274,8 @@ kms_ice_nice_agent_finalize (GObject * object)
 
   GST_DEBUG_OBJECT (self, "finalize");
 
+  g_main_context_iteration(self->priv->context, FALSE);
+
   g_clear_object (&self->priv->agent);
   g_slist_free_full (self->priv->remote_candidates, g_object_unref);
```

I cannot judge, though, whether this should be considered a bug in libnice or not.

**Comment by j1elo** [comment]
Good catch! Your solution is good for this issue, but it would hide yet another leak we seem to have and should fix too: the NiceAgent object (inherits the GObject class, which implements reference-counting semantics) has been referenced somewhere else, without a corresponding unref. Result is a dangling reference, so the ref count of the NiceAgent is 2 right before `g_clear_object (&self->priv->agent);` in kmsiceniceagent.c.

I could prove it because adding a printf in the `nice_agent_dispose()` function (in libnice), didn't print. Then added this code in kmsiceniceagent.c:

```
GObject *agentobj = G_OBJECT (self->priv->agent);
GST_WARNING_OBJECT (self, "finalize kms agent, ref count: %u", agentobj->ref_count);

// g_clear_object (&self->priv->agent);
while (agentobj->ref_count > 0) {
    g_object_unref (agentobj);
}
```

Then the printf started appearing in the output. My idea was to tackle this from this point of view: NiceAgent should be unref and released right when the KmsIceNiceAgent is released. This will run the `nice_agent_dispose()` function which releases all resources, including streams ([see here](https://gitlab.freedesktop.org/libnice/libnice/-/blob/a9c2846a89c4e74efbd7ef59fcfe7ff05efb2c5e/agent/agent.c#L5351)).

I've been reading on debugging reference issues with GObject, and sadly the tooling is practically nonexistent there. A tool named "refdbg" exists but it is unmaintained, and the best course of action seems to dive into runtime debugging with GDB, so I'll be doing that next week.

With the dangling reference issue fixed, we should see if the sockets are still leaked, in which case adding the `g_main_context_iteration()` will be good too.

**Comment by j1elo** [comment]
It turns out that the NiceAgent reference was being leaked because of changes that I had done myself in the process of studying the leaked sockets! So, the original proposal works fine and I'll incorporate it ASAP.

**Comment by ifplusor** [comment]
@j1elo Is this released in 6.13.2?

**Comment by j1elo** [comment]
On 11/5/20 10:38, James Yin wrote:
>
> @j1elo <https://github.com/j1elo> Is this released in 6.13.2?
>
yes

**Comment by ifplusor** [comment]
@j1elo I meet **Segmentation fault** after install 6.13.2, log in error.log is:

```
Segmentation fault (thread 140655612512000, pid 4437)
Stack trace:
[kms_ice_candidate_set_address]
/usr/lib/x86_64-linux-gnu/libkmswebrtcendpointlib.so.6:0x12CB5
[kms_webrtc_session_get_stream_id]
/usr/lib/x86_64-linux-gnu/libkmswebrtcendpointlib.so.6:0xEEBC
[g_closure_invoke]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x1010D
[g_signal_handler_disconnect]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x2305E
[g_signal_emit_valist]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x2B715
[g_signal_emit_by_name]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x2C608
[kms_ice_nice_agent_new_selected_pair_full]
/usr/lib/x86_64-linux-gnu/libkmswebrtcendpointlib.so.6:0x14888
[g_closure_invoke]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x1010D
[g_signal_handler_disconnect]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x2305E
[g_signal_emitv]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x2AC80
[nice_agent_close_async]
/usr/lib/x86_64-linux-gnu/libnice.so.10:0xDEB7
[nice_agent_gather_candidates]
/usr/lib/x86_64-linux-gnu/libnice.so.10:0x14D50
[kms_ice_nice_agent_get_type]
/usr/lib/x86_64-linux-gnu/libkmswebrtcendpointlib.so.6:0x1527B
[kms_webrtc_session_get_stream_id]
/usr/lib/x86_64-linux-gnu/libkmswebrtcendpointlib.so.6:0xF8E7
[__kms_webrtc_marshal_BOOLEAN__VOID]
/usr/lib/x86_64-linux-gnu/libkmswebrtcendpointlib.so.6:0x15AFE
[g_closure_invoke]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0:0x1010
```

and error in another log file:

```
error glib                      libnice:0 ()  file address.c: line 147 (nice_address_set_port): should not be reached
```
